node/network/statement-distribution/src/lib.rs

```diff
@@ -0,0 +1,1376 @@
+
+		let candidate = {
+			let mut c = AbridgedCandidateReceipt::default();
+			c.relay_parent = hash_c;
+			c.parachain_index = 1.into();
+			c
+		};
+		let candidate_hash = candidate.hash();
+
+		let old_view = View(vec![hash_a, hash_b]);
+		let new_view = View(vec![hash_b, hash_c]);
+
+		let mut active_heads = HashMap::new();
+		let validators = vec![
+			Sr25519Keyring::Alice.public().into(),
+			Sr25519Keyring::Bob.public().into(),
+			Sr25519Keyring::Charlie.public().into(),
+		];
+
+		let session_index = 1;
+		let signing_context = SigningContext {
+			parent_hash: hash_c,
+			session_index,
+		};
+
+		let new_head_data = {
+			let mut data = ActiveHeadData::new(validators, session_index);
+
+			let noted = data.note_statement(SignedFullStatement::sign(
+				Statement::Seconded(candidate.clone()),
+				&signing_context,
+				0,
+				&Sr25519Keyring::Alice.pair().into(),
+			));
+
+			assert_matches!(noted, NotedStatement::Fresh(_));
+
+			let noted = data.note_statement(SignedFullStatement::sign(
+				Statement::Valid(candidate_hash),
+				&signing_context,
+				1,
+				&Sr25519Keyring::Bob.pair().into(),
+			));
+
+			assert_matches!(noted, NotedStatement::Fresh(_));
+
+			let noted = data.note_statement(SignedFullStatement::sign(
+				Statement::Valid(candidate_hash),
+				&signing_context,
+				2,
+				&Sr25519Keyring::Charlie.pair().into(),
+			));
+
+			assert_matches!(noted, NotedStatement::Fresh(_));
+
+			data
+		};
+
+		active_heads.insert(hash_c, new_head_data);
+
+		let mut peer_data = PeerData {
+			view: old_view,
+			view_knowledge: {
+				let mut k = HashMap::new();
+
+				k.insert(hash_a, Default::default());
+				k.insert(hash_b, Default::default());
+
+				k
+			},
+		};
+
+		let pool = ThreadPool::new().unwrap();
+		let (mut ctx, mut handle) = subsystem_test::make_subsystem_context(pool);
+		let peer = PeerId::random();
+
+		executor::block_on(async move {
+			update_peer_view_and_send_unlocked(
+				peer.clone(),
+				&mut peer_data,
+				&mut ctx,
+				&active_heads,
+				new_view.clone(),
+			).await.unwrap();
+
+			assert_eq!(peer_data.view, new_view);
+			assert!(!peer_data.view_knowledge.contains_key(&hash_a));
+			assert!(peer_data.view_knowledge.contains_key(&hash_b));
+
+			let c_knowledge = peer_data.view_knowledge.get(&hash_c).unwrap();
+
+			assert!(c_knowledge.known_candidates.contains(&candidate_hash));
+			assert!(c_knowledge.sent_statements.contains(
+				&(CompactStatement::Candidate(candidate_hash), 0)
+			));
+			assert!(c_knowledge.sent_statements.contains(
+				&(CompactStatement::Valid(candidate_hash), 1)
+			));
+			assert!(c_knowledge.sent_statements.contains(
+				&(CompactStatement::Valid(candidate_hash), 2)
+			));
+
+			// now see if we got the 3 messages from the active head data.
+			let active_head = active_heads.get(&hash_c).unwrap();
+
+			// semi-fragile because hashmap iterator ordering is undefined, but in practice
+			// it will not change between runs of the program.
```

Comment:
we could consider using a `IndexMap` where iteration is done based on insertion order

Reply:
Great. Actually, this will allow a merge of `seconded_statement` and `other_statements`, since those are primarily the way they are for ensuring the iterator visits `Seconded` statements before others (which are dependent)